### ISSUE TITLE
Implement NCE loss to speed up Softmax with large amount of labels

### ISSUE THREAD

**Issue by xlvector** [dmlc/mxnet#2704]
For word based LSTM, the Softmax layer receive large amount of labels which make training very slow. Two methods can solve this problem: hierarchical softmax (HS) and NCE.

I think NCE is better than HS, so I want to implement NCE Loss similar to this repos: https://github.com/yandex/faster-rnnlm

The API may looks like

NCEOutput(data = input, label = target, noise = [negative samples])

Do you have more suggestions on this?

**Comment by sxjscience** [comment]
Tensorflow also has a nce_loss that we can refer to. https://www.tensorflow.org/versions/r0.9/api_docs/python/nn.html#nce_loss

**Comment by xlvector** [comment]
Thanks. Another problem I find is LSTM on words always cause OOM. However, NCE
should not affect memory.

How to I measure memory usage in SoftmaxOutput of LSTM with num_hidden, default_bucket_key and num_label. I think it should be num_hidden * num_label. But I find it seems LSTM use num_hidden * num_label * default_bucket_key

**Comment by piiswrong** [comment]
Is default_bucket_key = length? If you have output at each time step that would make sense

**Comment by xlvector** [comment]
yes. so if seq length is too long. it will cause oom？
i have vocab size 5000，hidden = 100，length=3000

**Comment by xlvector** [comment]
I think the NCE Loss operator should not replace Softmax, it should replace last FullConnect because we can not calculate nhidden * nlabel matrix in last FC.

So, NCE is FC + Softmax Loss, merge two op into single op. What's your ideas?

**Comment by xlvector** [comment]
I use following code. Here, label's shape is (batch_size, num_sampled + num_true)
`
      Tensor<xpu, 2, DType> data = in_data[nceloss::kData].get_with_shape<xpu, 2, DType>(                                                                                                                 
          Shape2(ishape[0], ishape.ProdShape(1, ishape.ndim())), s);                                                                                                                                      
      Tensor<xpu, 2, DType> label = in_data[nceloss::kLabel].get<xpu, 2, DType>(s);                                                                                                                       
      Tensor<xpu, 2, DType> wmat = in_data[nceloss::kWeight].get<xpu, 2, DType>(s);                                                                                                                       
      Tensor<xpu, 2, DType> out = out_data[nceloss::kOut].get<xpu, 2, DType>(s);                                                                                                                          
                                                                                                                                                                                       
      DType* cpu_labels = static_cast<DType*>(label.dptr_);                                                                                                                                               
      if (in_data[nceloss::kLabel].dev_mask_ == gpu::kDevMask) {                                                                                                                                          
  #if MXNET_USE_CUDA                                                                                                                                                                                      
        cpu_labels = reinterpret_cast<DType*>(malloc(label.MSize() * sizeof(DType)));                                                                                                                     
        cudaError_t cuda_status = cudaMemcpyAsync(cpu_labels, label.dptr_,                                                                                                                                
                                                  label.MSize() * sizeof(DType),                                                                                                                          
                                                  cudaMemcpyDeviceToHost,                                                                                                                                 
                                                  ctx.get_stream<gpu>()->stream_);                                                                                                                        
        CHECK_EQ(cuda_status, cudaSuccess) << "cuda memcpy label error";                                                                                                                                  
  #endif                                                                                                                                                                                                  
      }                                                                                                                                                                                                   
                                                                                                                                                                                                          
      for(int i = 0; i < label.size(0); i++) {                                                                                                                                                            
        for(int j = 0; j < label.size(1); j++) {                                                                                                                                                          
          Tensor<xpu, 1, DType> outij = out[i].Slice(j, j+1);                                                                                                                                             
         int la = (int)(cpu_labels[i * label.size(1) + j]);                                                                                                            
         VectorDot(outij, data[i], wmat[la]);                                                                                                                                                        
       }                                                                                                                                                                                                 
     }
`

However, the VectorDot seems does not work. Following are error message:

[18:38:58] /dl/xlvector/mxnet/dmlc-core/include/dmlc/logging.h:235: [18:38:58] /dl/xlvector/mxnet/mshadow/mshadow/./././dot_engine-inl.h:669: Check failed: dst.size(0) == sleft[0] && dst.size(1) == sright[1] && sleft[1] == sright[0] dot-gemm: matrix shape mismatch
[18:38:58] /dl/xlvector/mxnet/dmlc-core/include/dmlc/logging.h:235: [18:38:58] src/engine/./threaded_engine.h:306: [18:38:58] /dl/xlvector/mxnet/mshadow/mshadow/./././dot_engine-inl.h:669: Check failed: dst.size(0) == sleft[0] && dst.size(1) == sright[1] && sleft[1] == sright[0] dot-gemm: matrix shape mismatch
An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
terminate called after throwing an instance of 'dmlc::Error'
  what():  [18:38:58] src/engine/./threaded_engine.h:306: [18:38:58] /dl/xlvector/mxnet/mshadow/mshadow/./././dot_engine-inl.h:669: Check failed: dst.size(0) == sleft[0] && dst.size(1) == sright[1] && sleft[1] == sright[0] dot-gemm: matrix shape mismatch
An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
Aborted (core dumped)


The shape of data[i] is (100, )
The shape of wmat[la] is (100, )
The shape of outij is (1, )

**Comment by xlvector** [comment]
I try to write an C++ NceLoss layer to do this, but find its too slow.
However, after refer to tensorflow's implement, I find we can use existing op to do NCE loss.

So, I will submit an python example for this.

**Comment by sxjscience** [comment]
That's good. I'm planning to create another folder to include some OPs that can be constructed by basic c++ ops (Like LSTM). NCE can be one of them.

**Comment by sxjscience** [comment]
Closed by https://github.com/dmlc/mxnet/pull/2756
